src/fsharp/vs/IncrementalBuild.fs

```diff
@@ -1664,17 +1664,29 @@ type IncrementalBuilder(ctokCtor: CompilationThreadToken, frameworkTcImportsCach
     let MaxTimeStampInDependencies cache (ctok: CompilationThreadToken) (output:INode) = 
         IncrementalBuild.MaxTimeStampInDependencies cache ctok output.Name partialBuild 
 
-    member this.IncrementUsageCount() = 
-        assertNotDisposed() 
-        referenceCount  <- referenceCount  + 1
-        { new System.IDisposable with member x.Dispose() = this.DecrementUsageCount() }
+    /// Try to increment the usage count. If it returns None, then the object has been disposed by someone else already and may no longer be used. 
+    member this.TryIncrementUsageCount() =
+        let oldValue = referenceCount
+        if oldValue = 0 then None
+        else
+            let newValue = oldValue + 1
+            let exchanged = System.Threading.Interlocked.CompareExchange(&referenceCount, oldValue, newValue)
+            if exchanged = oldValue then
+                assertNotDisposed()
+                Some { new System.IDisposable with member x.Dispose() = this.DecrementUsageCount() }
+            else if exchanged = 0 then None
+            else
+                // someone else changed 'referenceCount' in the meantime - try again
+                this.TryIncrementUsageCount()
```

Comment:
It is probably over-complex. I was trying to account for a builder in a schrödinger state, where you have a reference to the object, but don't know if it is still alive.

Lets take a builder, which has been disposed, and the ref count is 0.

Two threads enter Increment. The first increments from 0 to 1, and correctly detects that the object is actually dead. BUT the second will then increment it from 1 to 2 and incorrectly revive a dead builder.

The loop is a standard way to deal with this, by reading the initial value, incrementing it, and using compare-exchange to set it. If another thread changed the value between reading it and updating it, the operation is retried.